alibi/explainers/kernel_shap.py

```diff
@@ -0,0 +1,780 @@
+                        self.ignore_weights = True
+
+            # NB: we have already summarised the data at this point
+            if self.summarise_background:
+
+                weights_dim = len(weights)
+                if background_data.ndim == 1:
+                    n_background_samples = 1
+                else:
+                    if not self.transposed:
+                        n_background_samples = background_data.shape[0]
+                    else:
+                        n_background_samples = background_data.shape[1]
+
+                if weights_dim != n_background_samples:
+                    msg = "The number of weights vector provided ({}) did not match the number of " \
+                          "summary data points ({}). The weights provided will be ignored!"
+                    logging.warning(msg.format(weights_dim, n_background_samples))
+
+                    self.ignore_weights = True
+
+    def _summarise_background(self,
+                              background_data: Union[shap.common.Data, pd.DataFrame, np.ndarray, sparse.spmatrix],
+                              n_background_samples: int) -> \
+            Union[shap.common.Data, pd.DataFrame, np.ndarray, sparse.spmatrix]:
+        """
+        Summarises the background data to n_background_samples in order to reduce the computational cost. If the
+        background data is a shap.common.Data object, no summarisation is performed.
+
+        Parameters
+        ----------
+            See constructor.
+
+        Returns
+        -------
+            If the user has specified grouping, then the input object is subsampled and and object of the same
+            type is returned. Otherwise, a shap.common.Data object containing the result of a kmeans algorithm
+            is wrapped in a shap.common.DenseData object and returned. The samples are weighted according to the
+            frequency of the occurence of the clusters in the original data.
+        """
+
+        if isinstance(background_data, shap.common.Data):
+            msg = "Received option to summarise the data but the background_data object " \
+                  "was an instance of shap.common.Data. No summarisation will take place!"
+            logging.warning(msg)
+            return background_data
+
+        if background_data.ndim == 1:
+            msg = "Received option to summarise the data but the background_data object only had " \
+                  "one record with {} features. No summarisation will take place!"
+            logging.warning(msg.format(len(background_data)))
+            return background_data
+
+        self.summarise_background = True
+
+        if isinstance(background_data, sparse.spmatrix):
+            return shap.sample(background_data, nsamples=n_background_samples)
+        elif self.use_groups or self.categorical_names:
+            return shap.sample(background_data, nsamples=n_background_samples)
```

Comment:
Yes. This is because I decided that passing a sparse array was an indication that there are categorical variables in the data so didn't want to use kmeans in that case. Initially I went for this layout to keep the code format tidy, can compress it a bit if you want.
```python
        # if the input is sparse, we assume there are categorical variables and use random sampling, not kmeans
        if self.use_groups or self.categorical_names or isinstance(background_data, sparse.spmatrix):
            return shap.sample(background_data, nsamples=n_background_samples)
        else:
            logging.info(
                "When summarising with kmeans, the samples are weighted in proportion to their "
                "cluster occurrence frequency. Please specify a different weighting of the samples "
                "through the by passing a weights of len=n_background_samples to the constructor!"
            )
            return shap.kmeans(background_data, n_background_samples)
```